src/main/java/io/anserini/util/TrainingDataGenerator.java

```diff
@@ -0,0 +1,360 @@
+    appender.start();
+
+    config.addAppender(appender);
+
+    // Adding reference to the appender
+    AppenderRef ref = AppenderRef.createAppenderRef(loggerName, null, null);
+    AppenderRef[] refs = new AppenderRef[]{ref};
+
+    LoggerConfig loggerConfig = LoggerConfig.createLogger("false", Level.TRACE,
+            loggerName, "true", refs, null, config, null);
+
+    // Adding appender to logger, and adding logger to context
+    loggerConfig.addAppender(appender, null, null);
+
+    config.addLogger(loggerName, loggerConfig);
+    ctx.updateLoggers();
+  }
+
+  private void initializeIndex() throws IOException {
+    if (kbIndexDirectory == null && args.index != null) {
+      LOG.info("Initializing KB index from directory: {}", args.index);
+      kbIndexDirectory = FSDirectory.open(Paths.get(args.index));
+    }
+  }
+
+  /**
+   * Gets index reader from the directory specified in the user parameters
+   *
+   * @return KB index reader
+   * @throws IOException on error
+   */
+  private IndexReader getKbIndexReader() throws IOException {
+    if (kbIndexReader == null) {
+      initializeIndex();
+      kbIndexReader = DirectoryReader.open(kbIndexDirectory);
+    }
+
+    return kbIndexReader;
+  }
+
+  /**
+   * @return index searcher
+   * @throws IOException on error
+   */
+  IndexSearcher getKbIndexSearcher() throws IOException {
+    if (kbIndexSearcher == null) {
+      kbIndexSearcher = new IndexSearcher(getKbIndexReader());
+    }
+
+    return kbIndexSearcher;
+  }
+
+  /**
+   * Get the current default index analyzer or create it.
+   * Use EnglishAnalyzer as the default.
+   *
+   * @return kb index analyzer
+   * @throws Exception
+   */
+  Analyzer getKbIndexAnalyzer() throws Exception {
+    if (kbIndexAnalyzer == null) {
+      kbIndexAnalyzer = new EnglishAnalyzer();
+    }
+
+    return kbIndexAnalyzer;
+  }
+
+  /**
+   * Helper function. Extracts value from literal that has a type
+   * (whether the type is a language for a string literal or a basic data type
+   * like date, int, etc.)
+   *
+   * @param literalString the string representation of the literal, including its type
+   * @return value of the literal
+   */
+  static String extractValueFromTypedLiteralString(String literalString) {
+    return NTriplesUtil.parseLiteral(literalString, valueFactory).stringValue();
+  }
+
+  /**
+   * Helper function. Converts freebase URI to freebase mention id
+   *
+   * @param freebaseUri freebase uri, similar to
+   * @return freebase mention id
+   */
+  static String freebaseUriToFreebaseId(String freebaseUri) {
+    return freebaseUri.substring(freebaseUri.lastIndexOf('/')).replace('.', '/');
+  }
+
+  /**
+   * Writes an array as a line separated by TRAINING_DATA_SEPARATOR.
+   *
+   * @param l    the logger to write to (the training data file)
+   * @param data array of data
+   */
+  void writeToTrainingFiles(Logger l, String... data) {
+    StringBuilder sb = new StringBuilder();
```

Comment:
No, `String...` allows for variable number of parameters without the need to create an array explicitly from the calling function.
It is translated to String[] anyway.